modules/core/src/main/scala/fs2/kafka/internal/KafkaConsumerActor.scala

```diff
@@ -175,44 +201,29 @@ private[kafka] final class KafkaConsumerActor[F[_], K, V](
       } else (st, F.unit)
     }
 
-    def removeRevokedRecords(revokedNonFetches: SortedSet[TopicPartition]) = withState { st =>
-      if (revokedNonFetches.nonEmpty) {
-        val revokedRecords = st.records.filterKeysStrict(revokedNonFetches)
-
-        if (revokedRecords.nonEmpty) {
-          val newState = st.withoutRecords(revokedRecords.keySet)
-
-          val action = logging.log(RemovedRevokedRecords(revokedRecords, newState))
-
-          (newState, action)
-        } else (st, F.unit)
-      } else (st, F.unit)
-    }
-
     ref
       .modify { state =>
-        val withRebalancing = state.withRebalancing(true)
+        val updatedState = state
+          .withRebalancing(true)
+          .withoutAssignments(revoked)
 
-        val fetches = withRebalancing.fetches.keySetStrict
-        val records = withRebalancing.records.keySetStrict
+        val fetches = updatedState.fetches.keySetStrict
+        val records = updatedState.records.keySetStrict
 
         val revokedFetches = revoked intersect fetches
-        val revokedNonFetches = revoked diff revokedFetches
 
         val withRecords = records intersect revokedFetches
         val withoutRecords = revokedFetches diff records
 
         (for {
           completeWithRecords <- completeWithRecords(withRecords)
           completeWithoutRecords <- completeWithoutRecords(withoutRecords)
-          removeRevokedRecords <- removeRevokedRecords(revokedNonFetches)
```

Comment:
Why don't you clean up the `record` container from records that belong to revoked partitions and have no downstream sinks?

Reply:
Described in above comment.
The whole fix is based on postponing removing records - we wait until assignment to make sure that we can revoke records. This is a small step handling Greedy rebalance protocols.